File: 15.3-sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#

# @lc code=start
from typing import List
from collections import defaultdict
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        nums.sort()
        ans = []

        for i in range(n):
            if i > 0 and nums[i - 1] == nums[i]:
                continue
            j = i + 1
            k = n - 1

            while j < k:
                s = nums[i] + nums[j] + nums[k]
                if s > 0:
                    k -= 1
                elif s < 0:
                    j += 1
                else:
                    ans.append([nums[i], nums[j], nums[k]])
                    j += 1

                    while j < k and nums[j] == nums[j - 1]:
                        j += 1
        return ans




        
# @lc code=end

File: 15.3-sum.java
/*
 * @lc app=leetcode id=15 lang=java
 *
 * [15] 3Sum
 */
import java.util.HashMap;
import java.util.Map;
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
// @lc code=start
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        //! 枚举 + 双指针
        //solve the duplication problem
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();
        int n = nums.length;
        

        for(int k = n -1; k > 0; k--){
            if(k < n - 1 && nums[k] == nums[k + 1]) continue;
            int i = 0;
            int j = k -1;
            while(i < j){
                int sum = nums[i] + nums[j] + nums[k];
                if(sum == 0){
                    ans.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    // 跳过与当前 i、j 相同的值，避免同一 k 下重复
                    int li = nums[i], lj = nums[j];
                    while (i < j && nums[i] == li) i++;
                    while (i < j && nums[j] == lj) j--;
                }
                else if(sum > 0){
                    j--;
                }else{
                    i++;
                }
            }
        }

        return ans;
    }
}
// @lc code=end

